app/controllers/mass_rollouts_controller.rb

```diff
@@ -47,34 +47,41 @@ def review_deploy
     end
   end
 
-  # TODO: show when deploys are waiting for buddy
-  # TODO: show when deploys were invalid
-  # TODO: show when stage had no reference
   def deploy
-    stages_to_deploy = Stage.find(params.require(:stage_ids))
-    deploy_references = stages_to_deploy.map do |stage|
+    warnings = []
+
+    stages = Stage.find(params[:stage_ids] || [])
+    deploys = stages.map do |stage|
       reference =
         case params[:reference_source]
         when 'template' then last_successful_template_reference(stage)
         when 'redeploy' then stage.last_successful_deploy&.reference
-        else
-          return unsupported_option(:reference_source)
+        else return unsupported_option(:reference_source)
         end
 
-      [stage, reference] if reference
+      unless reference
+        warnings << [stage, "No reference found"]
+        next
+      end
+
+      deploy = DeployService.new(current_user).deploy(stage, reference: reference)
+      if deploy.new_record?
```

Comment:
why not `unless deploy.valid?`

Reply:
that would run validations again ... we could check `errors` but that would not  catch things callbacks throwing ... only alternative is `id` but that's even less readable